### PR TITLE
[DRAFT] ESM - WebGL and its extensions

### PR DESCRIPTION
**Pull request by RaananW**
First - disclaimer:

This is a WIP to move the engine architecture to function-based esm compatible architecture. This PR addresses the thinengine/webgl engine and partly (!) webgpu. WebGPU is still being worked on, but i thought it would be important to already release this PR to get some reviews and questions.

Due to a few recent PRs (and an open one) I haven't merged master to this, so the engine functionality is a few weeks old. A future change to this PR will be to merge from master and update functions that were changed.

## How to use? 

@esm/core is ready to be consumed right after building it. To build run

`npm run build -w @esm/core`

For the time being, importing directly from esm/core will need to be done from its dist directory. Of course, this will not be a part of the final architecture.

After building you can install @esm/core using npm link or uding a direct file:// link alongside @babylonjs/core

Any function from the webgl engine is used independently using the engineState object you initially create. This state object holds all variables needed for the different functions to work. 

To use, for example:

```javascript
import { Color4 } from "@babylonjs/core/Maths/math.color";
import { initWebGLEngineState, clear, setViewport, createTexture } from "@esm/core/dist/Engines/WebGL/engine.webgl";

const canvas = document.getElementById("babylon-canvas") as HTMLCanvasElement;

const engineState = initWebGLEngineState(canvas);

setViewport(engineState, {
    x: 0,
    y: 0,
    width: 1,
    height: 1
});
clear(engineState, Color4.FromInts(83, 156, 33, 255), true, true);

createTexture(engineState, "./alien.png", false, false, null);
```

## Adapters

You can generate an adapter that will be "converted" to a corresponding Engine class. For example, you can generate a ThinEngine using this command:

```javascript
const engineState = initWebGLEngineState(glContext, {
         disableWebGL2Support: true
      });
      // get adapter
      const adapter: Engine = augmentEngineState(engineState, {
        // functions you are using in your experience, imported from the webgl engine

      });
```

Adapters are useful when you want to use ESM to construct a class that requires an engine in its constructor. There are a few helper/adapter objects that can be used to generate an engine that fits the class you are constructing. for example, to create an effect, do the following:

```javascript
import { initWebGLEngineState } from "@esm/core/dist/Engines/engine.webgl";

import { augmentEngineState } from "@esm/core/dist/Engines/engine.adapters";
import { effectWebGLAdapter } from "@esm/core/dist/Engines/WebGL/engine.adapterHelpers";

const canvas = document.getElementById("babylon-canvas") as HTMLCanvasElement;

const engineState = initWebGLEngineState(canvas);
// augment the engine state
const engineForEffect = augmentEngineState(engineState, effectWebGLAdapter);
// now engineForEffect can be used to construct an effect
```

This mechanism is (still) used internally.

## Extensions

Engine extensions are now not a part of the engine itself, but external functions. You can still use them to augment the state object for legacy reasons.

To use an extension:

```javascript
// import the functions needed 
import * as multiRenderExtension from "@esm/core/dist/Engines/WebGL/Extensions/multiRender/multiRender.webgl.js";
import {setExtension, EngineExtensions} from "@esm/core/dist/Engines/Extensions/engine.extensions.js";

// To let the engine use it internally (when needed by the engine) you can register he extension
setExtension(engineState, EngineExtensions.MULTI_RENDER, multiRenderExtension);

// You can use the import directly using the engine state, and now the engine can use it internally, if needed
```

## Bottom line

Just contact me if you have any questions. If something doesn't work - contact me as well! :-) 

This will stay draft for the team to comment about anything here.

### CODE REVIEW

File: packages/esm/core/src/Engines/WebGL/engine.adapterHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    setTexture,
+    setDepthStencilTexture,
+    setTextureArray,
+    setTextureFromPostProcess,
+    bindUniformBufferBase,
+    bindUniformBlock,
+};
+
+
+export const effectWebGPUAdapter = {
```
Comment (Popov72) [inline packages/esm/core/src/Engines/WebGL/engine.adapterHelpers.ts:59]: I think you forgot to remove this one from the file?

File: packages/esm/core/src/Engines/engine.adapters.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Augment an engineState object with methods to simulate a real engine object
+ *
+ * @param engineState the engineState object to augment
+ * @param injectedMethods The methods that will be injected to the engineState
+ * @param force Should we force re-injecting the methods
+ * @returns The engineState cased to the requested engine type
+ */
+export function augmentEngineState<T extends ThinEngine, E = BaseEngineMethods | WebGLEngineMethods | WebGPUEngineMethods>(
```
Comment (Popov72) [inline packages/esm/core/src/Engines/engine.adapters.ts:29]: Why do you need to inject methods in the engine state when (it seems!) you don't use them? When a method of the engine must be called, you are doing `method_engine(ps, ...)` with ps=engine state, and not `ps.method_engine(...)`? But maybe I missed something, I didn't review everything.

**Review by Popov72 — COMMENTED** [review]
That is a monster of a PR, congrats!

I only have a few comments to make, as I didn't go into detail, I just wanted to understand the general idea.

However, what I'm wondering, and I know we've already talked about this, is how you're going to automatically transform this code into a class-based implementation (the one we currently have) so that we only write the code once? Or maybe the transformation will go the other way, we'll keep writing code in the existing code base and there'll be a process to convert it to ESM. Either way, it seems that creating an automated process to manage this conversion is a real challenge...

File: packages/esm/core/readme.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+## Extensions
+
+Engine extensions are now not a part of the engine itself, but external functions. You can still use them to augment the state object for legacy reasons.
+
+To use an extension:
+
+```javascript
+// import the functions needed
+import * as multiRenderExtension from "@esm/core/dist/Engines/WebGL/Extensions/multiRender/multiRender.webgl.js";
+import { setExtension, EngineExtensions } from "@esm/core/dist/Engines/Extensions/engine.extensions.js";
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/esm/core/readme.md:78]: in esm you should, yes. we add them post-build to our es6 packages as well. Otherwise modern build frameworks can't work with babylon.
The convention is not 100% clear on whether or not it should be mjs (for module js) or js, but i use js and setting the package to be a module in package.json

File: packages/esm/core/src/Engines/WebGL/engine.adapterHelpers.ts (hunk 1 above)
Comment (RaananW) [reply packages/esm/core/src/Engines/WebGL/engine.adapterHelpers.ts:59]: yes! thanks. Was forgotten when i restructured the directories. thanks :-)

File: packages/esm/core/src/Engines/engine.adapters.ts (hunk 2 above)
Comment (RaananW) [reply packages/esm/core/src/Engines/engine.adapters.ts:29]: it is true, i don't use them in esm, but this way the engine state is turning to a full engine and can be used in other components that require an engine. This way effect, internal textures and so on can still work without changing the types they accept in the construtor
I have considered different ways to implement the "legacy"  support. performance-wise, this was the best way, as it doesn't involve accessing extra objects or arrays.

File: packages/esm/core/readme.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+```javascript
+const engineState = initWebGLEngineState(glContext, {
+         disableWebGL2Support: true
+      });
+      // get adapter
+      const adapter: Engine = augmentEngineState(engineState, {
+        // functions you are using in your experience, imported from the webgl engine
+
+      });
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/esm/core/readme.md:48]: You are right, this is the one catch here - if the functions are not imported, it is impossible to know what functions are needed to run a certain component, unless you dive into the code of this component, or run your experience and add each function that the engine is missing (which is how i got different projects to work with esm. Took roughly 10 minutes to do that, but it is, of course, not optimal!)
This is what i was trying to solve with the adapter helpers. It is a tree-shakable list of objects that contain the functions needed to run a specific component. so far I did it for the internal texture and effect, but I am planning on expanding this. 
Extensions, for example, are being added as a whole. So if you use setExtension, it will augment the engine's state object, making these functions available for other components.

**Review by carolhmj — COMMENTED** [review]
Just general questions and observations while I'm trying to understand the principles better

File: packages/esm/core/src/Engines/Extensions/query/query.base.ts [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+        __occlusionDataStorage: _OcclusionDataStorage;
+
+        /**
+         * Access property
+         * @internal
+         */
+        _occlusionDataStorage: _OcclusionDataStorage;
+
+        /**
+         * fes number indicates the number of allowed retries before stop the occlusion query, fes is useful if the occlusion query is taking long time before to the query result is retrieved, the query result indicates if the object is visible within the scene or not and based on that Babylon.Js engine decides to show or hide the object.
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/esm/core/src/Engines/Extensions/query/query.base.ts:158]: oh gosh. yes. Thanks :-)
```suggestion
         * this number indicates the number of allowed retries before stop the occlusion query, this is useful if the occlusion query is taking long time before to the query result is retrieved, the query result indicates if the object is visible within the scene or not and based on that Babylon.Js engine decides to show or hide the object.
```

File: packages/esm/core/src/Engines/Extensions/query/query.base.ts [hunk 6]
```diff
@@ -0,0 +1,363 @@
+import type { Nullable, int } from "@babylonjs/core/types.js";
+import type { Observer } from "@babylonjs/core/Misc/observable.js";
+import { PerfCounter } from "@babylonjs/core/Misc/perfCounter.js";
+import type { _TimeToken } from "@babylonjs/core/Instrumentation/timeToken.js";
+import type { IBaseEnginePublic } from "../../engine.base.js";
+import { AbstractMesh } from "@babylonjs/core/Meshes/abstractMesh.js";
+
+// TODO there is a breaking change here - returning engine state instead of engine
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/esm/core/src/Engines/Extensions/query/query.base.ts:8]: Yes and no. is ESM there is no issue with breaking changes. BUT when converting it back to UMD there might be a need for a proxy between the type that is returned and a ThinEngine, for example. One way of solving this might be to use ReturnType or some other Tyepscript helper, but I still have a few things to solve before fixing this.

The same issue can be seen in any observable that notifies an Engine. in ESM it will notify the state object but in UMD it will notify with a fully constructed engine. Though their types might be related, it is a totally different typescript type.

I hope it's clear :-) . I will be happy to explain it better

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14516/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14516/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14516/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14516/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14516/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14516/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14516/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14516/merge#BCU1XR#0

File: packages/esm/core/src/Engines/WebGL/Extensions/rawTexture/engine.rawTexture.webgl.ts [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+    }
+    const fes = engineState as WebGLEngineStateFull;
+    // Babylon's internalSizedFomat but gl's texImage2D internalFormat
+    const internalSizedFomat = _getRGBABufferInternalSizedFormat(fes, type, format, useSRGBBuffer);
+
+    // Babylon's internalFormat but gl's texImage2D format
+    const internalFormat = _getInternalFormat(fes, format);
+    const textureType = _getWebGLTextureType(fes, type);
+    _bindTextureDirectly(fes, fes._gl.TEXTURE_2D, texture, true);
+    _unpackFlipY(fes, invertY === undefined ? true : invertY ? true : false);
```
Comment (carolhmj) [inline packages/esm/core/src/Engines/WebGL/Extensions/rawTexture/engine.rawTexture.webgl.ts:43]: Why not `_unpackFlipY(fes, invertY === undefined ? true : invertY);`  or even `_unpackFlipY(fes, !!invertY);`?

File: packages/esm/core/src/Engines/WebGL/Extensions/rawTexture/engine.rawTexture.webgl.ts [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+        fes._gl.compressedTexImage2D(fes._gl.TEXTURE_2D, 0, (<any>fes._caps.s3tc)[compression], texture.width, texture.height, 0, <DataView>data);
+    } else {
+        fes._gl.texImage2D(fes._gl.TEXTURE_2D, 0, internalSizedFomat, texture.width, texture.height, 0, internalFormat, textureType, data);
+    }
+
+    if (texture.generateMipMaps) {
+        fes._gl.generateMipmap(fes._gl.TEXTURE_2D);
+    }
+    _bindTextureDirectly(fes, fes._gl.TEXTURE_2D, null);
+    //  this.resetTextureCache();
```
Comment (carolhmj) [inline packages/esm/core/src/Engines/WebGL/Extensions/rawTexture/engine.rawTexture.webgl.ts:67]: Missed comment? Or a todo to add?

File: packages/esm/core/src/Engines/WebGL/Extensions/rawTexture/engine.rawTexture.webgl.ts [hunk 9]
```diff
@@ -0,0 +1,597 @@
...
+                }
+            }
+
+            _bindTextureDirectly(fes, gl.TEXTURE_CUBE_MAP, null);
+        } else {
+            updateRawCubeTexture(fes, texture, faceDataArrays, format, type, invertY);
+        }
+
+        texture.isReady = true;
+        // this.resetTextureCache();
```
Comment (carolhmj) [inline packages/esm/core/src/Engines/WebGL/Extensions/rawTexture/engine.rawTexture.webgl.ts:358]: Same here

File: packages/esm/core/src/Engines/Extensions/transformFeedback/engine.transformFeedback.base.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
+import type { DataBuffer } from "@babylonjs/core/Buffers/dataBuffer.js";
+import type { Nullable } from "@babylonjs/core/types.js";
+import type { IBaseEnginePublic } from "../../engine.base.js";
+
+export const name = "transformFeedback";
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/esm/core/src/Engines/Extensions/transformFeedback/engine.transformFeedback.base.ts:5]: there is actually no real need for that. i'll clean it up

File: packages/esm/core/src/Engines/WebGL/Extensions/readTexture/readTexture.webgl.ts [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    switch (type) {
+        case Constants.TEXTURETYPE_BYTE: {
+            const buffer = sizeOrDstBuffer instanceof ArrayBuffer ? new Int8Array(sizeOrDstBuffer) : new Int8Array(sizeOrDstBuffer);
+            if (copyBuffer) {
+                buffer.set(new Int8Array(copyBuffer));
+            }
+            return buffer;
+        }
+        case Constants.TEXTURETYPE_UNSIGNED_BYTE: {
+            const buffer = sizeOrDstBuffer instanceof ArrayBuffer ? new Uint8Array(sizeOrDstBuffer) : new Uint8Array(sizeOrDstBuffer);
```
Comment (carolhmj) [inline packages/esm/core/src/Engines/WebGL/Extensions/readTexture/readTexture.webgl.ts:26]: Same here?

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14516/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by noname0310** [comment]
For use the WebGPUEngine with treeshaking in the esm package, I'm doing some side effect imports like this

```typescript
import "@babylonjs/core/Engines/WebGPU/Extensions/engine.alpha";
import "@babylonjs/core/Engines/WebGPU/Extensions/engine.rawTexture";
import "@babylonjs/core/Engines/WebGPU/Extensions/engine.renderTarget";
import "@babylonjs/core/Engines/WebGPU/Extensions/engine.uniformBuffer";
import { WebGPUEngine } from "@babylonjs/core/Engines/webgpuEngine";
```

If this PR is merged, will we have to do the same for the webGL engine in the future?

If so, I'm very happy that I can reduce the bundle size further.

**Comment by RaananW** [comment]
> For use the WebGPUEngine with treeshaking in the esm package, I'm doing some side effect imports like this
> 
> ```ts
> import "@babylonjs/core/Engines/WebGPU/Extensions/engine.alpha";
> import "@babylonjs/core/Engines/WebGPU/Extensions/engine.rawTexture";
> import "@babylonjs/core/Engines/WebGPU/Extensions/engine.renderTarget";
> import "@babylonjs/core/Engines/WebGPU/Extensions/engine.uniformBuffer";
> import { WebGPUEngine } from "@babylonjs/core/Engines/webgpuEngine";
> ```
> 
> If this PR is merged, will we have to do the same for the webGL engine in the future?
> 
> If so, I'm very happy that I can reduce the bundle size further.

In the long run, yes, probably. This will reduce package size. It will work differently in ESM - the state object will be augmented with the extension. So it's not exactly side effect if you use ESM directly, but will be a side effect if you are using the es6 package.

Sounds uber complicated, and I am sorry about that :-) I will fully document this feature once I am happy with the architecture.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14516/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14516/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14516/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14516/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14516/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by bjsplat** [comment]
Reviewer - this PR has made changes to one or more package.json files.

**Comment by RaananW** [comment]
Closing for the time being.
